Module4_Concepts/Depth-FirstSearch(DFS).md

```diff
@@ -0,0 +1,107 @@
+<!--title={Depth First Search (DFS)}-->
+
+# Depth-First Search 
+
+Depth-First Search(DFS) is one of our data searching algorithms we use in Python to navigate Data structures, Graphs, and Search Trees. DFS is best explained with a visual aide. 
+
+![]( https://upload.wikimedia.org/wikipedia/commons/thumb/1/1f/Depth-first-tree.svg/450px-Depth-first-tree.svg.png)
+
+DFS starts at a root point and branches on from that point. DFS continues down one path as far as possible before backtracking to the root point and taking an alternate path. 
+
+If we trace the numbers in the visual aide, from 1-12, we get a good sense of idea of how DFS can navigates data and graphs. 
+
+1. Choose one route and continue down that route until a dead end is reached. 
+
+   Circle 1 -> to Circle 2 -> to Circle 3 -> Circle 4
+
+2. Once a dead end is reached, we backtrack to previous nodes until we find another path not yet explored. 
+
+   Circle 4 -> to Circle 3 
+
+3. We continue down that new path. 
+
+   Circle 3 -> Circle 5 
+
+4. Then, we backtrack to the next available path. 
+
+   Circle 5 -> Circle 3 -> Circle 2
+
+5. Circle 2 has the available path of Circle 6.
+
+6. We continue this pattern until we have searched the entire data structure, tree, or graph. 
+
+To explain it in terms of python, we follow these rules.
+
+- **Rule 1** − Visit the adjacent unvisited vertex. Mark it as visited. Display it. Push it in a stack.
```

Comment:
why a stack? what is the significance of the vertices that are stored in the stack?